libminifi/src/utils/ThreadPool.cpp

```diff
@@ -0,0 +1,238 @@
+      // I'm very sorry for this - committee must has been seriously drunk when the interface of prio queue was submitted.
+      Worker<T> task = std::move(const_cast<Worker<T>&>(delayed_worker_queue_.top()));
+      delayed_worker_queue_.pop();
+      worker_queue_.enqueue(std::move(task));
+      tasks_available_.notify_one();
+    }
+    if(delayed_worker_queue_.empty()) {
+      delayed_task_available_.wait(lock);
+    } else {
+      auto wait_time = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - delayed_worker_queue_.top().getTimeSlice());
+      delayed_task_available_.wait_for(lock, wait_time);
+    }
+  }
+}
+
+template<typename T>
+bool ThreadPool<T>::execute(Worker<T> &&task, std::future<T> &future) {
+  {
+    std::unique_lock<std::mutex> lock(worker_queue_mutex_);
+    task_status_[task.getIdentifier()] = true;
+  }
+  future = std::move(task.getPromise()->get_future());
+  bool enqueued = worker_queue_.enqueue(std::move(task));
+  if (running_) {
+    tasks_available_.notify_one();
+  }
+
+  task_count_++;
+
+  return enqueued;
+}
+
+template<typename T>
+void ThreadPool<T>::manageWorkers() {
+  for (int i = 0; i < max_worker_threads_; i++) {
+    std::stringstream thread_name;
+    thread_name << name_ << " #" << i;
+    auto worker_thread = std::make_shared<WorkerThread>(thread_name.str());
+    worker_thread->thread_ = createThread(std::bind(&ThreadPool::run_tasks, this, worker_thread));
+    thread_queue_.push_back(worker_thread);
+    current_workers_++;
+  }
+
+  if (daemon_threads_) {
+    for (auto &thread : thread_queue_) {
+      thread->thread_.detach();
+    }
+  }
+
+// likely don't have a thread manager
+  if (LIKELY(nullptr != thread_manager_)) {
+    while (running_) {
+      auto waitperiod = std::chrono::milliseconds(500);
+      {
+        std::lock_guard<std::recursive_mutex> lock(manager_mutex_);
+        if (thread_manager_->isAboveMax(current_workers_)) {
+          auto max = thread_manager_->getMaxConcurrentTasks();
+          auto differential = current_workers_ - max;
+          thread_reduction_count_ += differential;
+        } else if (thread_manager_->shouldReduce()) {
+          if (current_workers_ > 1)
+            thread_reduction_count_++;
+          thread_manager_->reduce();
+        } else if (thread_manager_->canIncrease() && max_worker_threads_ > current_workers_) {  // increase slowly
+          std::unique_lock<std::mutex> lock(worker_queue_mutex_);
+          auto worker_thread = std::make_shared<WorkerThread>();
+          worker_thread->thread_ = createThread(std::bind(&ThreadPool::run_tasks, this, worker_thread));
+          if (daemon_threads_) {
+            worker_thread->thread_.detach();
+          }
+          thread_queue_.push_back(worker_thread);
+          current_workers_++;
+        }
+      }
+      {
+        std::lock_guard<std::recursive_mutex> lock(manager_mutex_);
```

Comment:
Why do we release and grab the same mutex again? It's much less overhead to just keep the mutex locked, except if there's waiting involved.

Reply:
Fair point.

Reply:
Ah, sorry, I missed the lines. In this case you are right, merged the blocks to grab the mutex only once.